io: SslTlsSupport throws assertion errors when spray client doesn't trust server certificates

See https://groups.google.com/d/topic/spray-user/lkwLsA2RJd8/discussion

```
59:46 DEBUG [.client.HttpHostConnectionSlot] - Dispatching HEAD request to https://localhost:8095/foo across connection Actor[akka://nio-ats/user/IO-HTTP/group-4/0#958044922]
59:46 ERROR [torRefProvider(akka://nio-ats)] - guardian failed, shutting down system
java.lang.AssertionError: assertion failed
        at scala.Predef$.assert(Predef.scala:165) ~[org.scala-lang.scala-library_2.10.3.v20130923-060037-VFINAL-e2fec6b28d.jar:na]
        at spray.io.SslTlsSupport$$anon$1$$anon$2$$anon$11.apply(SslTlsSupport.scala:283) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.SslTlsSupport$$anon$1$$anon$2$PumpAction.apply(SslTlsSupport.scala:246) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.SslTlsSupport$$anon$1$$anon$2$$anon$6$$anonfun$2.apply(SslTlsSupport.scala:85) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.SslTlsSupport$$anon$1$$anon$2$$anon$6$$anonfun$2.apply(SslTlsSupport.scala:81) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.SslTlsSupport$$anon$1$$anon$2.process(SslTlsSupport.scala:234) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.DynamicPipelines$$anonfun$eventPipeline$1.apply(Pipelines.scala:48) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.DynamicPipelines$$anonfun$eventPipeline$1.apply(Pipelines.scala:48) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.RawPipelineStage$$anon$3$$anonfun$2.apply(Pipelines.scala:115) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.RawPipelineStage$$anon$3$$anonfun$2.apply(Pipelines.scala:115) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.TickGenerator$$anon$1$$anon$2$$anonfun$1.apply(TickGenerator.scala:41) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.TickGenerator$$anon$1$$anon$2$$anonfun$1.apply(TickGenerator.scala:38) ~[io.spray.io_1.2.0.20131023.jar:na]
        at spray.io.ConnectionHandler$$anonfun$running$1.applyOrElse(ConnectionHandler.scala:58) ~[io.spray.io_1.2.0.20131023.jar:na]
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165) ~[org.scala-lang.scala-library_2.10.3.v20130923-060037-VFINAL-e2fec6b28d.jar:na]
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:498) [com.typesafe.akka.osgi_2.2.3.jar:na]
        at akka.actor.ActorCell.invoke(ActorCell.scala:456) [com.typesafe.akka.osgi_2.2.3.jar:na]
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:237) [com.typesafe.akka.osgi_2.2.3.jar:na]
        at akka.dispatch.Mailbox.run(Mailbox.scala:219) [com.typesafe.akka.osgi_2.2.3.jar:na]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110) [na:1.7.0_06]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603) [na:1.7.0_06]
        at java.lang.Thread.run(Thread.java:722) [na:1.7.0_06]
```

--- a/spray-io-tests/src/test/scala/spray/io/SslTlsSupportSpec.scala
+++ b/spray-io-tests/src/test/scala/spray/io/SslTlsSupportSpec.scala
@@ -230,6 +230,22 @@ class SslTlsSupportSpec extends Specification with NoTimeConversions {
       TestUtils.verifyActorTermination(clientConn.handler)
       server.close()
     }
+
+    "handle untrusted server certificates gracefully" in new TestSetup {
+      override implicit val sslContext = SSLContext.getDefault
+
+      val server = new SpraySslServer
+      val connAttempt = attemptSpraySslClientConnection(server.address)
+      val serverConn = server.acceptOne()
+      val clientConn = connAttempt.finishConnect()
+
+      clientConn.writeLn("Foo")
+      serverConn.events.expectMsg(Tcp.Aborted)
+      TestUtils.verifyActorTermination(serverConn.handler)
+      clientConn.events.expectMsgType[Tcp.ErrorClosed]
+      TestUtils.verifyActorTermination(clientConn.handler)
+      server.close()
+    }
   }
 
   step { system.shutdown() }

--- a/spray-io/src/main/scala/spray/io/SslTlsSupport.scala
+++ b/spray-io/src/main/scala/spray/io/SslTlsSupport.scala
@@ -89,7 +89,7 @@ object SslTlsSupport {
                     if (isOutboundDone) finishingClose(closedEvent)
                     else waitingForAck(remainingOutgoingData, closedEvent)
                   }
-                } else assert(!isOutboundDone)
+                } else verify(!isOutboundDone)
               case Tcp.PeerClosed     ⇒ receivedUnexpectedPeerClosed()
               case x: Tcp.ErrorClosed ⇒ eventPL(x) // is there anything we need to close in this case?
               case x @ (_: Tcp.ConnectionClosed | WriteChunkAck) ⇒
@@ -253,7 +253,6 @@ object SslTlsSupport {
           // pumping is only stopped by a buffer underflow on the inbound side or when both sides are done
           val encrypt: PumpAction = new PumpAction {
             @tailrec def apply(tempBuf: ByteBuffer): Unit = {
-              assert(!engine.isOutboundDone)
               tempBuf.clear()
               val result = engine.wrap(pendingOutboundBytes, tempBuf)
               tempBuf.flip()
@@ -278,7 +277,6 @@ object SslTlsSupport {
           // same as `encrypt` but starts with decrypting
           val decrypt: PumpAction = new PumpAction {
             @tailrec def apply(tempBuf: ByteBuffer): Unit = {
-              assert(!engine.isInboundDone)
               tempBuf.clear()
               val result = engine.unwrap(pendingInboundBytes, tempBuf)
               tempBuf.flip()
@@ -319,7 +317,7 @@ object SslTlsSupport {
           def encryptedBytesPending = pendingEncryptedBytes.length > 0 // why doesn't ByteStringBuilder have an `isEmpty`?
 
           def setPendingOutboundBytes(buffer: ByteBuffer): Unit = {
-            assert(!bytesLeft(pendingOutboundBytes))
+            verify(!bytesLeft(pendingOutboundBytes))
             pendingOutboundBytes = buffer
           }
 
@@ -345,7 +343,7 @@ object SslTlsSupport {
             if (engine.isInboundDone) {
               // our pumping logic should make sure that we immediately outbound close and flush
               // after having detected an inbound close
-              assert(engine.isOutboundDone)
+              verify(engine.isOutboundDone)
               true
             } else engine.isOutboundDone
 
@@ -358,6 +356,9 @@ object SslTlsSupport {
             log.debug("Failing write command because " + msg)
             context.self ! cmd.failureMessage
           }
+
+          def verify(condition: Boolean): Unit =
+            if (!condition) throw new IllegalStateException
         }
 
       def writeChunkStream(cmd: Tcp.WriteCommand): Stream[WriteChunk] = {

